src/main/java/com/iexec/worker/chain/IexecHubService.java

```diff
@@ -175,6 +177,65 @@ private IexecHubContract.TaskRevealEventResponse sendRevealTransaction(String ch
         return null;
     }
 
+    public Optional<ChainReceipt> contributeAndFinalize(Contribution contribution, String resultDigest, String resultLink, String callbackData) {
+        try {
+            return CompletableFuture.supplyAsync(() -> {
+                log.info("contributeAndFinalize request [chainTaskId:{}, waitingTxCount:{}]",
+                        contribution.getChainTaskId(), getWaitingTransactionCount());
+                IexecHubContract.TaskFinalizeEventResponse finalizeEvent = sendContributeAndFinalizeTransaction(contribution, resultDigest, resultLink, callbackData);
+                return Optional.ofNullable(finalizeEvent)
+                        .map(event -> ChainUtils.buildChainReceipt(event.log, contribution.getChainTaskId(), getLatestBlockNumber()));
+            }, executor).get();
+        } catch (ExecutionException e) {
+            log.error("contributeAndFinalize asynchronous execution did not complete", e);
+        } catch (InterruptedException e) {
+            log.error("contributeAndFinalize thread has been interrupted", e);
+            Thread.currentThread().interrupt();
+        }
+        return Optional.empty();
+    }
+
+    private IexecHubContract.TaskFinalizeEventResponse sendContributeAndFinalizeTransaction(Contribution contribution, String resultDigest, String resultLink, String callbackData) {
```

Comment:
Not sure why you don't get `resultDigest` from `contribution` :thinking:

Reply:
because it is a miss, fixing this

Reply:
please see cd8d77c